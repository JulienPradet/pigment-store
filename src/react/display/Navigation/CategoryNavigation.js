import React from 'react'
import {Match, Link} from 'react-router'
import Highlighter from 'react-highlight-words'
import {featureContainsSearch, componentContainsSearch, categoryContainsSearch} from './index'
import {Container, Item} from '../util/View/SidebarMenu'

const FeatureNavigation = ({feature, pathname, search}) => (
  <Item fade>
    <Link to={pathname}>
      <Highlighter
        searchWords={[search]}
        textToHighlight={feature.name}
      />
    </Link>
  </Item>
)

const ComponentSubNavigation = ({component, pathname, search, displayAll}) => (
  <Container>
    {component.features
      .filter((feature) => displayAll || search === '' || featureContainsSearch(search)(feature))
      .map((feature) => (
        <FeatureNavigation
          key={feature.name}
          pathname={pathname}
          feature={feature}
          search={search}
        />
      ))}
  </Container>
)

const ComponentNavigation = ({pathname, component, search}) => (
  <Match pattern={`${pathname}/component-${component.name}`} children={({matched, pattern}) => (
    <Item isActive={matched}>
      <Link to={pattern}>
        <Highlighter
          searchWords={[search]}
          textToHighlight={component.name}
        />
      </Link>
      {matched || componentContainsSearch(search)(component)
        ? <ComponentSubNavigation
          component={component}
          pathname={pattern}
          search={search}
          displayAll={matched}
        />
        : null
      }
    </Item>
  )} />
)

const CategorySubNavigation = ({category, pathname, search, displayAll}) => <Container>
  {category.categories
    .filter((category) => displayAll || search === '' || categoryContainsSearch(search)(category))
    .map((category) => (
      <CategoryNavigation
        key={category.name}
        category={category}
        pathname={pathname}
        search={search}
      />
    ))}
  {category.components
    .filter((component) => displayAll || search === '' || componentContainsSearch(search)(component))
    .map((component) => (
      <ComponentNavigation
        key={component.name}
        pathname={pathname}
        component={component}
        search={search}
      />
    ))}
</Container>

const CategoryNavigation = ({category, pathname, search}) => (
  <Match pattern={`${pathname}/category-${category.name}`} children={({matched, pattern}) => {
    return (
      <Item isActive={matched}>
        <Link to={pattern}>
          <Highlighter
            searchWords={[search]}
            textToHighlight={category.name}
          />
        </Link>
        {matched || categoryContainsSearch(search)(category)
          ? <CategorySubNavigation
            pathname={pattern}
            category={category}
            search={search}
            displayAll={matched}
          />
          : null
        }
      </Item>
    )
  }} />
)

export { CategorySubNavigation }
