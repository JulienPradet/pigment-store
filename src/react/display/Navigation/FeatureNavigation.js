import React from 'react'
import Highlighter from 'react-highlight-words'
import {compose, withProps} from 'recompose'

import {Item} from '../util/View/SidebarMenu'
import {ModalLink} from '../util/View/Modal'
import {makeFeatureView} from '../router'

export default compose(
  withProps(({pathPrefix, name}) => ({
    path: makeFeatureView(pathPrefix, name)
  })),
  withProps(({isActive, path}) => ({
    active: isActive(path)
  }))
)(({name, component, search, path, active}) => {
  return <Item isActive={active}>
    <ModalLink to={path}>
      <Highlighter
        searchWords={[search]}
        textToHighlight={name}
      />
  </ModalLink>
  </Item>
})
