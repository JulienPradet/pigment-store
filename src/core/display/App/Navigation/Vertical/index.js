import React from 'react'
import {Link} from 'react-router'
import {compose, withState, withHandlers} from 'recompose'
import {MenuTitle, Item, Search} from '../../util/View/SidebarMenu'
import {CategorySubNavigation} from './CategoryNavigation'

export const isMatching = (search, name) => search !== '' && name.match(new RegExp(search, 'i'))

export const featureContainsSearch = (search) => feature => {
  return isMatching(search, feature.name)
}

export const componentContainsSearch = (search) => (component) => {
  return isMatching(search, component.name) ||
    component.features.some(featureContainsSearch(search))
}

export const categoryContainsSearch = (search) => (category) => {
  return isMatching(search, category.name) ||
      category.components.some(componentContainsSearch(search)) ||
      category.categories.some(categoryContainsSearch(search))
}

const Navigation = ({search, onSearchChange, indexCategory, isActive, prefix = ''}) => <div>
  <MenuTitle><Link to='/'>Pigment Store</Link></MenuTitle>
  <Item>
    <Search search={search} onChange={onSearchChange} />
  </Item>
  <Item>
    <CategorySubNavigation category={indexCategory} pathname={prefix} search={search} />
  </Item>
</div>

const SmartNavigation = compose(
  withState('search', 'setSearch', ''),
  withHandlers({
    onSearchChange: ({setSearch}) => (search) => setSearch(search)
  })
)(Navigation)

SmartNavigation.displayName = 'Navigation'

export default SmartNavigation
