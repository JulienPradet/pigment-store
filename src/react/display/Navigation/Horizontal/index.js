import React from 'react'
import CategoryNavigation from './CategoryNavigation'

const Navigation = ({indexCategory, prefix}) => (
  <CategoryNavigation category={indexCategory} prefix={prefix} parentPattern={prefix} />
)

export default Navigation
