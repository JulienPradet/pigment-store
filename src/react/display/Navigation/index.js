import React from 'react'
import {Link} from 'react-router'
import {compose, withState, withHandlers} from 'recompose'
import {MenuTitle, Item, Search} from '../util/View/SidebarMenu'
import {CategorySubNavigation} from './CategoryNavigation'

export const isMatching = (search, name) => name.match(new RegExp(search, 'i'))

export const featureContainsSearch = (search) => feature => {
  return isMatching(search, feature.name)
}

export const componentContainsSearch = (search) => (component) => {
  return isMatching(search, component.name) ||
    Object.keys(component.features)
      .map((key) => component.features[key])
      .some(featureContainsSearch(search))
}

export const categoryContainsSearch = (search) => ({name, category}) => {
  return isMatching(search, name) ||
    category.components.some(componentContainsSearch(search)) ||
    category.categories.some(categoryContainsSearch(search))
}

const Navigation = ({search, onSearchChange, indexCategory, isActive}) => <div>
  <MenuTitle><Link to='/'>Pigment Store</Link></MenuTitle>
  <Item>
    <Search search={search} onChange={onSearchChange} />
  </Item>
  <CategorySubNavigation category={indexCategory} pathPrefix='' search={search} isActive={isActive} />
</div>

export default compose(
  withState('search', 'setSearch', ''),
  withHandlers({
    onSearchChange: ({setSearch}) => (search) => setSearch(search),
    isActive: ({location}) => (path) => location.pathname.startsWith(path) || location.pathname.startsWith('/preview' + path)
  })
)(Navigation)
