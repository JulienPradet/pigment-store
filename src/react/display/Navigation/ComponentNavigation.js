import React from 'react'
import Highlighter from 'react-highlight-words'
import {compose, withProps} from 'recompose'
import {Link} from 'react-router'

import {isMatching, featureContainsSearch} from './index'
import FeatureNavigation from './FeatureNavigation'
import {Container, Item} from '../util/View/SidebarMenu'
import {componentNameToPath} from '../router'

const ComponentSubNavigation = ({component, pathPrefix, search, displayAll, isActive}) => <Container>
  {Object.keys(component.features)
    .map((key) => component.features[key])
    .filter((feature) => displayAll || featureContainsSearch(search)(feature))
    .map((feature) => <FeatureNavigation
      key={feature.name}
      pathPrefix={pathPrefix}
      name={feature.name}
      feature={feature}
      search={search}
      isActive={isActive}
    />)}
</Container>

export default compose(
  withProps(({pathPrefix, name}) => ({
    path: `${pathPrefix}/${componentNameToPath(name)}`
  })),
  withProps(({isActive, path}) => ({
    active: isActive(path)
  }))
)(({name, component, search, path, active, isActive, displayAll}) => {
  return <Item isActive={active}>
    <Link to={path}>
      <Highlighter
        searchWords={[search]}
        textToHighlight={name}
      />
    </Link>
    {active
      ? <ComponentSubNavigation
        component={component}
        pathPrefix={path}
        search={search}
        displayAll={displayAll || isMatching(search, name)}
        isActive={isActive}
      />
      : null}
  </Item>
})
