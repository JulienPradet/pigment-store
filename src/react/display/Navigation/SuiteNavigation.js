import React from 'react'
import {compose, withProps} from 'recompose'
import {Link} from 'react-router'
import Highlighter from 'react-highlight-words'

import {isMatching, componentContainsSearch} from './index'
import ComponentNavigation from './ComponentNavigation'
import {Container, Item} from '../util/View/SidebarMenu'
import {makePath} from '../router'

const SuiteSubNavigation = ({suite, pathPrefix, search, displayAll, isActive}) => <Container>
  {Object.keys(suite.components)
    .map((key) => suite.components[key])
    .filter((component) => displayAll || componentContainsSearch(search)(component))
    .map((component) => <ComponentNavigation
      key={component.name}
      pathPrefix={pathPrefix}
      name={component.name}
      component={component}
      search={search}
      displayAll={displayAll}
      isActive={isActive}
    />)}
</Container>

export default compose(
  withProps(({pathPrefix, suite}) => ({
    path: makePath(pathPrefix, suite.name)
  })),
  withProps(({isActive, path}) => ({
    active: isActive(path)
  }))
)(({suite, search, path, active, isActive, displayAll}) => {
  return <Item isActive={active}>
    <Link to={path}>
      <Highlighter
        searchWords={[search]}
        textToHighlight={suite.name}
      />
    </Link>
    {active
      ? <SuiteSubNavigation
        suite={suite}
        pathPrefix={path}
        search={search}
        displayAll={displayAll || isMatching(search, suite.name)}
        isActive={isActive}
      />
      : null}
  </Item>
})
