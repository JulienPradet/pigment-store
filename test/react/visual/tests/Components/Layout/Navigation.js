import React from 'react'
import {MemoryRouter} from 'react-router'
import PigmentStore from '../../../../../../src/index'
import Navigation from '../../../../../../src/core/display/App/Navigation/Vertical/index'
import indexCategory from '../fixtures/index'

export default PigmentStore.React.describe('Navigation', Navigation)
  .setDescription(`
The left navigation that allows to explore the styleguide
  `)
  .featureJsx(
    'Default',
    (
      <MemoryRouter>
        <Navigation indexCategory={indexCategory} />
      </MemoryRouter>
    )
  )
  .featureJsx(
    'With an active category',
    (
      <MemoryRouter location='/category-Atom'>
        <Navigation indexCategory={indexCategory} />
      </MemoryRouter>
    )
  )
  .featureJsx(
    'With an active component',
    (
      <MemoryRouter location='/category-Atom/component-Avatar'>
        <Navigation indexCategory={indexCategory} />
      </MemoryRouter>
    )
  )
