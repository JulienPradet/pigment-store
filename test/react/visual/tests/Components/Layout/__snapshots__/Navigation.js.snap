exports[`Navigation Default 1`] = `
<div>
  <h2
    className="title">
    <a
      className=""
      href="/"
      onClick={[Function]}
      style={Object {}}>
      Pigment Store
    </a>
  </h2>
  <li
    className="item">
    <div
      className="search">
      <input
        onChange={[Function]}
        placeholder="Search..."
        type="text"
        value="" />
      <button
        onClick={[Function]}>
        <i
          className="material-icons">
          search
        </i>
      </button>
    </div>
  </li>
  <li
    className="item">
    <ul
      className="list">
      <li
        className="item">
        <a
          className=""
          href="/category-Atom"
          onClick={[Function]}
          style={Object {}}>
          <span>
            <span>
              Atom
            </span>
          </span>
        </a>
      </li>
      <li
        className="item">
        <a
          className=""
          href="/category-Molecules"
          onClick={[Function]}
          style={Object {}}>
          <span>
            <span>
              Molecules
            </span>
          </span>
        </a>
      </li>
    </ul>
  </li>
</div>
`;

exports[`Navigation With an active category 1`] = `
<div>
  <h2
    className="title">
    <a
      className=""
      href="/"
      onClick={[Function]}
      style={Object {}}>
      Pigment Store
    </a>
  </h2>
  <li
    className="item">
    <div
      className="search">
      <input
        onChange={[Function]}
        placeholder="Search..."
        type="text"
        value="" />
      <button
        onClick={[Function]}>
        <i
          className="material-icons">
          search
        </i>
      </button>
    </div>
  </li>
  <li
    className="item">
    <ul
      className="list">
      <li
        className="item activeItem">
        <a
          className=""
          href="/category-Atom"
          onClick={[Function]}
          style={Object {}}>
          <span>
            <span>
              Atom
            </span>
          </span>
        </a>
        <ul
          className="list">
          <li
            className="item">
            <a
              className=""
              href="/category-Atom/component-Title"
              onClick={[Function]}
              style={Object {}}>
              <span>
                <span>
                  Title
                </span>
              </span>
            </a>
          </li>
          <li
            className="item">
            <a
              className=""
              href="/category-Atom/component-Avatar"
              onClick={[Function]}
              style={Object {}}>
              <span>
                <span>
                  Avatar
                </span>
              </span>
            </a>
          </li>
        </ul>
      </li>
      <li
        className="item">
        <a
          className=""
          href="/category-Molecules"
          onClick={[Function]}
          style={Object {}}>
          <span>
            <span>
              Molecules
            </span>
          </span>
        </a>
      </li>
    </ul>
  </li>
</div>
`;

exports[`Navigation With an active component 1`] = `
<div>
  <h2
    className="title">
    <a
      className=""
      href="/"
      onClick={[Function]}
      style={Object {}}>
      Pigment Store
    </a>
  </h2>
  <li
    className="item">
    <div
      className="search">
      <input
        onChange={[Function]}
        placeholder="Search..."
        type="text"
        value="" />
      <button
        onClick={[Function]}>
        <i
          className="material-icons">
          search
        </i>
      </button>
    </div>
  </li>
  <li
    className="item">
    <ul
      className="list">
      <li
        className="item activeItem">
        <a
          className=""
          href="/category-Atom"
          onClick={[Function]}
          style={Object {}}>
          <span>
            <span>
              Atom
            </span>
          </span>
        </a>
        <ul
          className="list">
          <li
            className="item">
            <a
              className=""
              href="/category-Atom/component-Title"
              onClick={[Function]}
              style={Object {}}>
              <span>
                <span>
                  Title
                </span>
              </span>
            </a>
          </li>
          <li
            className="item activeItem">
            <a
              className=""
              href="/category-Atom/component-Avatar"
              onClick={[Function]}
              style={Object {}}>
              <span>
                <span>
                  Avatar
                </span>
              </span>
            </a>
            <ul
              className="list">
              <li
                className="item fade">
                <span>
                  <span>
                    Default
                  </span>
                </span>
              </li>
              <li
                className="item fade">
                <span>
                  <span>
                    Placeholder
                  </span>
                </span>
              </li>
            </ul>
          </li>
        </ul>
      </li>
      <li
        className="item">
        <a
          className=""
          href="/category-Molecules"
          onClick={[Function]}
          style={Object {}}>
          <span>
            <span>
              Molecules
            </span>
          </span>
        </a>
      </li>
    </ul>
  </li>
</div>
`;
